File: src/test/java/org/hisp/dhis/rules/models/RuleTests.java
package org.hisp.dhis.rules.models;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import static org.assertj.core.api.Java6Assertions.assertThat;
import static org.mockito.Mockito.mock;

@RunWith( JUnit4.class )
public class RuleTests
{

    @Test( expected = NullPointerException.class )
    public void createShouldThrowOnNullCondition()
    {
        Rule.create( "test_program_stage", 1, null, new ArrayList<RuleAction>(), "", "" );
    }

    @Test( expected = NullPointerException.class )
    public void createShouldThrowOnNullActionsList()
    {
        Rule.create( "test_program_stage", 1, "test_condition", null, "", "" );
    }

    @Test
    public void createShouldPropagatePropertiesCorrectly()
    {
        RuleAction ruleAction = mock( RuleAction.class );

        Rule rule = Rule.create( "test_program_stage", 1,
            "test_condition", Arrays.asList( ruleAction ), "", "uid" );

        assertThat( rule.programStage() ).isEqualTo( "test_program_stage" );
        assertThat( rule.condition() ).isEqualTo( "test_condition" );
        assertThat( rule.priority() ).isEqualTo( 1 );
        assertThat( rule.actions().size() ).isEqualTo( 1 );
        assertThat( rule.actions().get( 0 ) ).isEqualTo( ruleAction );
        assertThat( rule.uid() ).isEqualTo( "uid" );
    }

    @Test( expected = UnsupportedOperationException.class )
    public void createShouldReturnImmutableList()
    {
        RuleAction ruleActionOne = mock( RuleAction.class );
        RuleAction ruleActionTwo = mock( RuleAction.class );

        List<RuleAction> actions = new ArrayList<>();
        actions.add( ruleActionOne );
        actions.add( ruleActionTwo );

        Rule rule = Rule.create( "test_program_stage", 1, "test_condition", actions, "", "" );

        // mutating source array
        actions.clear();

        assertThat( rule.actions().size() ).isEqualTo( 2 );
        assertThat( rule.actions().get( 0 ) ).isEqualTo( ruleActionOne );
        assertThat( rule.actions().get( 1 ) ).isEqualTo( ruleActionTwo );

        rule.actions().clear();
    }
}

File: src/test/java/org/hisp/dhis/rules/models/RuleTests.kt
package org.hisp.dhis.rules.models

import kotlinx.collections.immutable.persistentListOf
import org.assertj.core.api.Java6Assertions.assertThat
import org.junit.Test
import org.junit.rules.ExpectedException
import org.junit.runner.RunWith
import org.junit.runners.JUnit4
import org.mockito.Mockito.mock
import kotlin.test.assertFailsWith

@RunWith(JUnit4::class)
class RuleTests {

    val thrown: ExpectedException = ExpectedException.none()

    @Test
    fun createShouldThrowOnNullCondition() {
        assertFailsWith<NullPointerException> {
            Rule.create("test_program_stage", 1, null!!, persistentListOf(), "")
        }
    }

    @Test
    fun createShouldThrowOnNullActionsList() {
        assertFailsWith<NullPointerException> {
            Rule.create("test_program_stage", 1, "test_condition", null!!, "")
        }
    }

    @Test
    fun createShouldPropagatePropertiesCorrectly() {
        val ruleAction = mock(RuleAction::class.java)
        val rule = Rule.create("test_program_stage", 1,
                "test_condition", persistentListOf(ruleAction), "")

        assertThat(rule.programStage).isEqualTo("test_program_stage")
        assertThat(rule.condition).isEqualTo("test_condition")
        assertThat(rule.priority).isEqualTo(1)
        assertThat(rule.actions.size).isEqualTo(1)
        assertThat(rule.actions[0]).isEqualTo(ruleAction)
    }

    @Test
    fun createShouldReturnImmutableList() {
        val ruleActionOne = mock(RuleAction::class.java)
        val ruleActionTwo = mock(RuleAction::class.java)

        val actions = persistentListOf(ruleActionOne, ruleActionTwo)
        val rule = Rule.create("test_program_stage", 1, "test_condition", actions, "")

        actions.clear()

        assertThat(rule.actions.size).isEqualTo(2)
        assertThat(rule.actions[0]).isEqualTo(ruleActionOne)
        assertThat(rule.actions[1]).isEqualTo(ruleActionTwo)

        // tests immutability
        assertThat(actions.size).isEqualTo(2)
    }
}
